### ISSUE TITLE
New feature: setting the initial state when instantiating the state machine

### ISSUE THREAD

**Issue by wizfromoz** [jp-embedded/scxmlcc#103]
I'm trying to implement a persistent state machine. One of the requirements is on powerup the state machine goes straight into the last saved state, without any on-entry actions being invoked. Looking into the implementation, I can't see how this can be done with scxmlcc generated state machine directly.

I can think of work-arounds, but it would greatly simplify things if the **state machine constructor could accept the initial state as a parameter**.

**Comment by sstiller** [comment]
That's not in the SCXML standard.
Is it somehow related to #23?

> constructor could accept the initial state as a parameter

That could become interesting with parallel (sub)states.

**Comment by wizfromoz** [comment]
I wasn't aware of the <history> element, but after reading about it, that's precisely that. There's an event (going into hibernation) that forces the state to be recorded and another event that should should restore into that state. The specification says "SCXML processor must record ...". I suppose that the state machine implementation must provide hooks that do the actual persisting? Anyway, is this on the table for consideration?

I was looking at using **new_state<T>()** to set the initial state by directly assigning it to the model.cur_state. The states that are in the leaves of the state tree would have an action that, on receiving **save** event would persist their id. Then, on *restore* event, I would set the current state to the stored one. Any reason why this couldn't work?

**Comment by jp-embedded** [comment]
It has not been at the table before now.

With a persistant state machine, I guess the data model should be stored as well.

**Comment by jp-embedded** [comment]
Also as sstiler mentions, we should be careful about features that does not follow the standard. 

Maybe this persistence could be implemented by making a class that inherits the state machine. Then you can overload the init method at set your initial state there - just a quick thought. I am on vacation rigtht now 😊

**Comment by wizfromoz** [comment]
Thanks Jan, I like the idea about overloading *init* method. It will do for now, until you get around implementing <history> tag, which *is* standard, and would help with persisting the state machine.

**Comment by jp-embedded** [comment]
Good to hear.

I am currently working on another issue with handling conflicting transitions in parallel states 

But eventually, the history tag will be implemented.
